src/out_of_order_engine.cc

```diff
@@ -0,0 +1,414 @@
+
+	target_state host_queue_target_state;
+	dense_map<device_id, target_state> device_queue_target_states;
+
+	/// The set of all instructions between submit() and complete_assigned(). Keyed by `instruction_id` to allow collecting successors through iterating over a
+	/// newly submitted instruction's dependencies, which are given in terms of instruction ids. Any dependency that is not found in `incomplete_instructions`
+	/// is assumed to have completed earlier (triggering its removal from the map).
+	std::unordered_map<instruction_id, incomplete_instruction_state> incomplete_instructions;
+
+	/// Queue of all instructions in `eagerly_assignable_state` and `immediately_assignable_state`, in decreasing order of instruction priority.
+	std::priority_queue<const instruction*, std::vector<const instruction*>, instruction_priority_less> assignment_queue;
+
+	explicit engine_impl(const system_info& system);
+
+	target_state& get_target_state(const target tgt, const std::optional<device_id>& device);
+
+	/// Retrieve state for an existing lane.
+	lane_state& get_lane_state(const target tgt, const std::optional<device_id>& device, const lane_id lane);
+
+	/// Linearly search for a lane that has no in-flight instructions within a target_state. If none exists, add an additional lane.
+	lane_id get_free_lane_id(const target tgt, const std::optional<device_id>& device);
+
+	/// Attempt to replace assignment state for an incomplete instruction with `eagerly_assignable_state` or `immediately_assignable_state` in response to
+	/// either the initial submission or to completion of one or more predecessors. If successful, the instruction ends up in `assignment_queue`.
+	void try_mark_for_assignment(incomplete_instruction_state& node);
+
+	void submit(const instruction* const instr);
+
+	void complete(const instruction* const instr);
+
+	bool is_idle() const;
+
+	/// Return the highest-priority instruction from `assignment_queue` that is unconditionally assignable (helper).
+	incomplete_instruction_state* pop_assignable();
+
+	std::optional<assignment> assign_one();
+};
+
+engine_impl::engine_impl(const system_info& system) : system(system), device_queue_target_states(system.devices.size()) {}
+
+target_state& engine_impl::get_target_state(const target tgt, const std::optional<device_id>& device) {
+	switch(tgt) {
+	case target::host_queue: assert(!device.has_value()); return host_queue_target_state;
+	case target::device_queue: assert(device.has_value()); return device_queue_target_states[*device];
+	default: utils::unreachable();
+	}
+}
+
+lane_state& engine_impl::get_lane_state(const target tgt, const std::optional<device_id>& device, const lane_id lane) {
+	return get_target_state(tgt, device).lanes.at(lane);
+}
+
+lane_id engine_impl::get_free_lane_id(const target tgt, const std::optional<device_id>& device) {
+	auto& target_state = get_target_state(tgt, device);
+	for(lane_id lid = 0; lid < target_state.lanes.size(); ++lid) {
```

Comment:
I assume we expect the number of lanes per target to remain rather small, so doing a linear search should be fine. But how small exactly, and how do we ensure that?